### PR TITLE
Overall code reformat, README.md reformat, and use of AssemblyPublicizer to kill off ugly Reflection code (and preview.png upd'd!)

### PR DESCRIPTION
**Pull request by StupidRepo**
# What's changed?
- Format the code to look nicer
  - Also changed position of Harmony attributes
- Made use of BepInEx's AssemblyPublicizer to reduce annoying code which can otherwise be prevented by using the publicizer.
- ~~Use `.csproj.user` for user variables.~~
  - ~~Changed `.gitignore` to exclude this .user file from being added by Git so others don't leak something (it's just a file path until it's not just a file path 😨)~~
- Updated `readme.md` to reflect changes (also renamed to uppercase - `README.md`)
- Updated `preview.png` to look... well... yeah. It is 256x256 but **whatever displays it should already be configured to downscale it to fit anyway** (at least I hope so!).
  - Can we get the game to please show mod icons though?

Love the game, and hopefully you accept my PR! :P

**P.S.: I did build the project and everything was working like it usually would.**

### CODE REVIEW

**Comment by b1bradders** [comment]
Should I also put the patch & settings class in a seperate folder & file or no?

**Comment by C0mputery** [comment]
Using `.csproj.user` for user variables is obtuse and and less clear than the existing condition. This would lead to new developers having a harder time building projects.

**Comment by Hamunii** [comment]
> Using `.csproj.user` for user variables is obtuse and and less clear than the existing condition. This would lead to new developers having a harder time building projects.

I agree that having everything in one file can feel nice at least in theory. But it's terrible for contributors since they need to make a PR which implements this anyways. Also, I think it's good to have a template that does things well so people learn to do things well. Otherwise you might not learn that there's a better way.

**Comment by b1bradders** [comment]
> > Using `.csproj.user` for user variables is obtuse and and less clear than the existing condition. This would lead to new developers having a harder time building projects.
> 
> I agree that having everything in one file can feel nice at least in theory. But it's terrible for contributors since they need to make a PR which implements this anyways. Also, I think it's good to have a template that does things well so people learn to do things well. Otherwise you might not learn that there's a better way.

Can't tell if you're agreeing with Computery or not. Regardless, the reason for `.csproj.user` is the following:

1) easier to update file paths than assigning env variables (i mean, going through those intellij/vs settings/system settings to find the env variables will get annoying at some point)
2) not pushed to git (even tho env variables aren't but still)
3) it uses standard csproj layout, where .csproj.user starts with `<Project></Project>` and you can define anything in there.

**Comment by khyperia** [comment]
Thanks for the PR! While I agree that the `.csproj.user` is nicer when everyone knows what's going on, unfortunately, this repo is aimed at people just being introduced to modding. Everything should be as clear, obvious, and simple as possible, which means doing things the nice way isn't the best way. Please revert that change.

> Should I also put the patch & settings class in a seperate folder & file or no?

Keep them in the same file! Again, things should be as straightforward as possible.

**Comment by RedBigz** [comment]
> Keep them in the same file! Again, things should be as straightforward as possible.

@khyperia I feel like a foldered structure should play nicely with most IDEs, like with my mods I usually make separate classes for each patch to avoid scrolling through one massive file, and it should be great for newcomers to learn how to organise their project structure, but if it's a simple fix I can get behind putting your mod all in one plugin file.

I think AssemblyPublicizer in MSBuild is an awesome addition, so newcomers don't get baffled around reflection and write unnecessary code.

The Harmony attributes look better fitted to the code in terms of style, but you should clarify that the class attribute's type **must** match the one defined in `nameof` or else Reflection can't discover any functions.

For the most part, it looks decently good to me.

**Comment by b1bradders** [comment]
> Thanks for the PR! While I agree that the `.csproj.user` is nicer when everyone knows what's going on, unfortunately, this repo is aimed at people just being introduced to modding. Everything should be as clear, obvious, and simple as possible, which means doing things the nice way isn't the best way. Please revert that change.

Alright :)

**Comment by b1bradders** [comment]
> The Harmony attributes look better fitted to the code in terms of style, but you should clarify that the class attribute's type **must** match the one defined in `nameof` or else Reflection can't discover any functions.

Yeah, I will add some comments in a sec. `nameof` simply takes a method so VSCode/Rider/other IDEs should be able to autocomplete methods accordingly. I like `nameof` for this reason instead of using a raw string lol. If a function name is changed in a game update, look! It's gone red! So, it just gives me that visual feedback 😉

File: README.md [hunk 1]
```diff
@@ -7,10 +7,11 @@ This is an example mod for content warning to show how to use the modding API!
 
 ## This project
 
-- ExampleCWPlugin.csproj: this specifies the location of Content Warning and its dlls. If it's not installed in the default steam location of `C:\Program Files (x86)\Steam\steamapps\common\Content Warning`, you can edit that here, or set the environment variable CWDir before opening the project.
-  - There's also a build step that copies the built dll and preview.png to the content warning plugins directory when building, letting you easily test your code by just building then launching content warning.
-- preview.png: This is the image used for your mod's steam workshop page when uploading the mod.
-- Example.cs: The code for the mod. Check out the comments in the code for details!
+- `ExampleCWPlugin.csproj`: This points MSBuild/the IDE to the location of Content Warning and its DLLs. If it's not installed in the default Steam location of `C:\Program Files (x86)\Steam\steamapps\common\Content Warning`, you can edit that in the `.csproj.user` file, or set the environment variable `CWDir` before opening the project.
```
Comment (khyperia) [inline README.md]: revert the change talking about `csproj.user`, here and the added line below

File: .gitignore [hunk 2]
```diff
@@ -1,4 +1,5 @@
 .idea
 bin
 obj
-*.DotSettings.user
\ No newline at end of file
+*.DotSettings.user
+*.user
```
Comment (khyperia) [inline .gitignore]: can revert this change to .gitignore

File: ExampleCWPlugin.csproj [hunk 3]
```diff
@@ -10,10 +10,19 @@
...
+    <!-- Use BepInEx's Assembly Publicizer to tell the compiler & IDE that every field, method, etc. is public, in the game assemblies. -->
+    <!-- This makes it easier to change fields and call methods that would otherwise be private. -->
+    <ItemGroup>
+      <PackageReference Include="BepInEx.AssemblyPublicizer.MSBuild" Version="0.4.2">
+        <PrivateAssets>all</PrivateAssets>
+        <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+      </PackageReference>
+    </ItemGroup>
+
+    <Target Name="CopyOutputFileToGame" AfterTargets="AfterBuild">
```
Comment (khyperia) [inline ExampleCWPlugin.csproj]: is there a reason you changed this from `AfterTargets="Build"` to `AfterTargets="AfterBuild"`?

**Review by khyperia — CHANGES_REQUESTED** [review]
almost there!

File: ExampleCWPlugin.csproj (hunk 3 above)
Comment (b1bradders) [reply ExampleCWPlugin.csproj]: Yes, for some reason Build would not trigger the target. All the times I've used Targets to copy something, I use AfterBuild and that has always worked.

Maybe it's IDE dependant? Hopefully not. Either way, this change was made just to fix this problem. :)

Comment (khyperia) [reply ExampleCWPlugin.csproj]: From what I can tell, [Build is now the preference instead of BeforeBuild/AfterBuild](https://github.com/dotnet/msbuild/issues/1680). Both Rider and `dotnet build` invoke the target for me. Probably best to go with the recommended style and fix whatever's broken with your build system!

Comment (b1bradders) [reply ExampleCWPlugin.csproj]: > From what I can tell, [Build is now the preference instead of BeforeBuild/AfterBuild](https://github.com/dotnet/msbuild/issues/1680). Both Rider and `dotnet build` invoke the target for me. Probably best to go with the recommended style and fix whatever's broken with your build system!

Oh, weird.... I use Rider too and I still had the problem. I'll try again in a minute with it set to Build :)

Comment (b1bradders) [reply ExampleCWPlugin.csproj]: Yep, that worked!
